compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InlineCallableReferenceToLambda.kt

```diff
@@ -49,191 +45,213 @@ internal val inlineCallableReferenceToLambdaPhase = makeIrFilePhase(
 //
 //      foo(::smth) -> foo { a -> smth(a) }
 //
-internal class InlineCallableReferenceToLambdaPhase(val context: JvmBackendContext) : FileLoweringPass {
-    private val inlinableCR = mutableSetOf<IrCallableReference>()
+internal class InlineCallableReferenceToLambdaPhase(val context: JvmBackendContext) : FileLoweringPass,
+    IrElementTransformerVoidWithContext() {
+
+    private val inlinableFunctionReferences = mutableSetOf<IrFunctionReference>()
+    private val inlinablePropertyReferences = mutableSetOf<IrPropertyReference>()
+
 
     override fun lower(irFile: IrFile) {
-        irFile.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
-
-            override fun visitFunctionAccess(expression: IrFunctionAccessExpression): IrExpression {
-                val callee = expression.symbol.owner
-                if (callee.isInlineFunctionCall(context)) {
-                    for (valueParameter in callee.valueParameters) {
-                        if (valueParameter.isInlineParameter()) {
-                            expression.getValueArgument(valueParameter.index)?.let { argument ->
-                                if (argument is IrCallableReference && isInlineIrExpression(argument)) {
-                                    inlinableCR.add(argument)
-                                }
-                            }
-                        }
-                    }
-                }
+        IrInlineReferenceLocator.scan(context, irFile).apply {
+            inlinableFunctionReferences.addAll(inlineFunctionReferences.filter { !lambdaToCallSite.containsKey(it.symbol.owner) })
```

Comment:
@kandersen Could you clarify this filtering?

I also can't catch the difference with inlinablePropertyReferences.addAll(inlinePropertyReferences), actualy property references could be passed to crossinline argument

Reply:
I can try! 

This pass should not "eta-expand" the function reference in a lambda in inline position. The IrInlineReferenceLocator tracks such lambdas in that map, and this was a convenient way to reuse that tracking. I can perhaps introduce a property on the Locator class that does the lookup in the keyset with a more meaningful name?